File: algorithm/merge_sort.cpp
MergeSort is a Divide and Conquer algorithm. 
It divides input array in two halves, calls itself for the two halves and then merges 
the two sorted halves. The merg() function is used for merging two halves. The merge(arr, l, m, r)
is key process that assumes that arr[l..m] and arr[m+1..r] are sorted and merges the two sorted sub-arrays into one. 
See following C implementation for details.

基本思想： divde and conquer
MergeSort(arr[], l,  r)
If r > l
     1. Find the middle point to divide the array into two halves:  
             middle m = (l+r)/2
     2. Call mergeSort for first half:   
             Call mergeSort(arr, l, m)
     3. Call mergeSort for second half:
             Call mergeSort(arr, m+1, r)
     4. Merge the two halves sorted in step 2 and 3:
             Call merge(arr, l, m, r)
             
  
/* C program for merge sort */
#include<stdlib.h>
#include<stdio.h>
 
/* Function to merge the two haves arr[l..m] and arr[m+1..r] of array arr[] */
void merge(int arr[], int l, int m, int r)
//doing the divide job
//l=0, m=(l+r)/2, r=sizeofarray-1;

{
    int i, j, k;
    int n1 = m - l + 1;//前半段的大小
    int n2 =  r - m;//后半段的大小
 
    /* create temp arrays */
    int L[n1], R[n2];
 
    /* Copy data to temp arrays L[] and R[] */
    for(i = 0; i < n1; i++)
        L[i] = arr[l + i];
    for(j = 0; j < n2; j++)
        R[j] = arr[m + 1+ j];//分别copy进两个不同的数组
 
    /* Merge the temp arrays back into arr[l..r]*/
    i = 0;
    j = 0;
    k = l;
    while (i < n1 && j < n2)
    {
        if (L[i] <= R[j])
        {
            arr[k] = L[i];
            i++;
        }
        else
        {
            arr[k] = R[j];
            j++;
        }
        k++;
    }//合并成一个数组，合并数组算法常用，可借鉴这里 
 
    /* Copy the remaining elements of L[], if there are any */
    while (i < n1)
    {
        arr[k] = L[i];
        i++;
        k++;
    }
 
    /* Copy the remaining elements of R[], if there are any */
    while (j < n2)
    {
        arr[k] = R[j];
        j++;
        k++;
    }
}
 
/* l is for left index and r is right index of the sub-array
  of arr to be sorted */
void mergeSort(int arr[], int l, int r)
{
    if (l < r)
    {
        int m = l+(r-l)/2; //Same as (l+r)/2, but avoids overflow for large l and h
        mergeSort(arr, l, m);
        mergeSort(arr, m+1, r);
        merge(arr, l, m, r);
    }
}
 
 
/* UITLITY FUNCTIONS */
/* Function to print an array */
void printArray(int A[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", A[i]);
    printf("\n");
}
 
/* Driver program to test above functions */
int main()
{
    int arr[] = {12, 11, 13, 5, 6, 7};
    int arr_size = sizeof(arr)/sizeof(arr[0]);
 
    printf("Given array is \n");
    printArray(arr, arr_size);
 
    mergeSort(arr, 0, arr_size - 1);
 
    printf("\nSorted array is \n");
    printArray(arr, arr_size);
    return 0;
}

重要思想：
一个功能， 两个任务：divide and conquer
分配给两个不同的function 来完成

Time Complexity: Sorting arrays on different machines.
Merge Sort is a recursive algorithm and time complexity can be expressed as following recurrence relation.
T(n) = 2T(n/2) + \Theta(n)
The above recurrence can be solved either using Recurrence Tree method or Master method. 
It falls in case II of Master Method and solution of the recurrence is \Theta(nLogn).
Time complexity of Merge Sort is \Theta(nLogn) in all 3 cases (worst, average and best) 
as merge sort always divides the array in two halves and take linear time to merge two halves.
Auxiliary Space: O(n)
Algorithmic Paradigm: Divide and Conquer
Sorting In Place: No in a typical implementation
Stable: Yes

File: sort/mergeSort.java

	public static void merge(int[] array, int l, int m, int r){
		int i, j, k;
		int n1 = m-l+1;
		int n2 = r-m;
		
		int[] L = new int[n1];
		int[] R = new int[n2];
		for(i = 0; i<n1; i++){
			L[i] = array[l+i];
		}
		for(i=0; i<n2; i++){
			R[i] = array[m+1+i];
		}
		
		/*Merge the temp arrays back into arr[l...r]*/
		i = 0;//initial index of first subarray
		j = 0;//initial index of second subarray
		k = l;//initial index of merged subarray
		while(i<n1 && j<n2){
			if(L[i]<R[j]){
				array[k] = L[i];
				i++;
			}else{
				array[k]= R[j];
				j++;
			}
			
			k++;
		}
		while(i<n1){
			array[k] = L[i];
			i++;
			k++;
		}
		while(j<n2){
			array[k] = R[j];
			j++;
			k++;
		}
	}
	
	public static void mergeSort(int[] array, int l, int r){
		if(l<r){
			int m = l+(r-l)/2;
			mergeSort(array, l, m);
			mergeSort(array, m+1, r);
			merge(array, l, m, r);
		}
	}
	
	public static void main(String []args){
		int[] arr = {9,8,7,6,5,4,3,2,1,0};
		mergeSort(arr, 0, arr.length-1);
		for(int i=0; i<arr.length; i++){
			System.out.println(arr[i]+" ");
		}
		
		
		
	}
